examples/all-clusters-app/all-clusters-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -0,0 +1,771 @@
+     *  kPluggedInDemand
+     */
+    switch (state)
+    {
+    case StateEnum::kNotPluggedIn:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+    case StateEnum::kPluggedInNoDemand:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+    case StateEnum::kPluggedInDemand:
+        // TODO - work out logic here
+        mHwState = state;
+        break;
+
+    default:
+        /* All other states should be managed by the Delegate */
+        // TODO (assert?)
+        break;
+    }
+
+    return Status::Success;
+}
+
+/**
+ * @brief    Called by EVSE Hardware to indicate a fault
+ *
+ * @param    FaultStateEnum
+ */
+Status EnergyEvseDelegate::HwSetFault(FaultStateEnum fault)
+{
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::Fault()");
+
+    if (fault == FaultStateEnum::kNoError)
+    {
+        /* Update State to previous state */
+        // TODO: need to work out the logic here!
+
+        /* Update SupplyState to previous state */
+    }
+    else
+    {
+        /* Update State & SupplyState */
+        SetState(StateEnum::kFault);
+        SetSupplyState(SupplyStateEnum::kDisabledError);
+    }
+
+    /* Update FaultState */
+    SetFaultState(fault);
+
+    // TODO: Generate events
+
+    return Status::Success;
+}
+
+/**
+ * @brief    Called by EVSE Hardware to Send a RFID event
+ *
+ * @param    ByteSpan RFID tag value (max 10 octets)
+ */
+Status EnergyEvseDelegate::HwSetRFID(ByteSpan uid)
+{
+    Events::Rfid::Type event{ .uid = uid };
+    EventNumber eventNumber;
+    CHIP_ERROR error = LogEvent(event, mEndpointId, eventNumber);
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "[Notify] Unable to send notify event: %s [endpointId=%d]", error.AsString(), mEndpointId);
+        return Status::Failure;
+    }
+
+    return Status::Success;
+}
+/**
+ * @brief    Called by EVSE Hardware to share the VehicleID
+ *
+ * This routine will make a copy of the string so the callee doesn't
+ * have to hold onto it forever.
+ *
+ * @param    CharSpan containing up to 32 chars.
+ */
+Status EnergyEvseDelegate::HwSetVehicleID(const CharSpan & newValue)
+{
+    if (mVehicleID.IsNull() || !newValue.data_equal(mVehicleID.Value()))
+    {
+        /* create a copy of the string so the callee doesn't have to keep it */
+        char * destinationBuffer = new char[kMaxVehicleIDBufSize];
+
+        MutableCharSpan destinationString(destinationBuffer, kMaxVehicleIDBufSize);
+        CHIP_ERROR err = CopyCharSpanToMutableCharSpan(newValue, destinationString);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(AppServer, "HwSetVehicleID - could not copy vehicleID");
+            delete[] destinationBuffer;
+            return Status::Failure;
+        }
+
+        if (!mVehicleID.IsNull())
+        {
+            delete[] mVehicleID.Value().data();
+        }
+
+        mVehicleID = MakeNullable(static_cast<CharSpan>(destinationString));
```

Comment:
will this own the buffer and handle releases? if this is a span it is generally not owning. Does this leak?

Reply:
See that this is deleted above when re-set. Please add a delete in the destructor too (and overall this needs refactoring, I understand #30993 is for that)